packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,164 @@
+ *
+ * 1. Filter the Performance Resource Timings by the name (it should match the
+ * XHR URL), additionally, the start/end timings of every performance entry
+ * should fit within the span start/end timings. These filtered performance
+ * resource entries are considered as possible entries associated to the xhr.
+ * Those are possible as there might be more than two entries that pass the
+ * filter.
+ *
+ * 2. As the XHR could cause a CORS pre-flight, we have to look for either
+ * possible pairs of performance resource timings or a single performance
+ * resource entry (a possible pair is when a resource timing entry does not
+ * overlap timings with other resource timing entry. Also, a possible single
+ * resource timing is when that resource timing entry is not paired with any
+ * other entry). Thus, for this step traverse the array of possible resource
+ * entries and for every entry try to pair it with the other possible entries.
+ *
+ * 3. Pick the best performance resource timing for the XHR: Using the possible
+ * performance resource timing entries from previous step, the best entry will
+ * be the one with the minimum gap to the span start/end timings. That is the
+ * substraction between the entry `respondeEnd` value and the span
+ * `endPerfTime` plus the substraction between the entry `startTime` and span
+ * `startPerfTime`. In case it is a tuple, the `startTime` corresponds to the
+ * first entry and the `responseEnd` is from second entry.
+ * The performance resource timing entry with the minimum gap to the span
+ * start/end timings points out that entry is the best fit for the span.
+ *
+ * @param xhrUrl
+ * @param span
+ */
+export function getXhrPerfomanceData(
+  xhrUrl: string,
+  span: Span
+): XhrPerformanceResourceTiming | undefined {
+  const filteredPerfEntries = getPerfResourceEntries(xhrUrl, span);
+  const possibleEntries = getPossiblePerfResourceEntries(filteredPerfEntries);
+  const bestEntry = getBestPerfResourceTiming(possibleEntries, span);
+  return bestEntry;
+}
+
+// Get Performance Resource Timings and filter them by matching the XHR url
+// with perfomance entry name. Additionally, the entry's start/end
+// timings must fit with in the span's start/end timings.
+export function getPerfResourceEntries(
+  xhrUrl: string,
+  span: Span
+): PerformanceResourceTiming[] {
+  return performance
+    .getEntriesByType('resource')
+    .filter(entry =>
+      isPerfEntryPartOfXhr(entry as PerformanceResourceTiming, xhrUrl, span)
+    ) as PerformanceResourceTiming[];
+}
+
+export function getPossiblePerfResourceEntries(
+  perfEntries: PerformanceResourceTiming[]
+): XhrPerformanceResourceTiming[] {
+  const possiblePerfEntries = new Array<XhrPerformanceResourceTiming>();
+  const pairedEntries = new Set<PerformanceResourceTiming>();
+  let perfEntry1: PerformanceResourceTiming;
+  let perfEntry2: PerformanceResourceTiming;
+  // As this part of the algorithm traverses the array twice, although,
```

Comment:
Would it be possible to use sorting to find the non-overlapping intervals more efficiently? See http://web.stanford.edu/class/archive/cs/cs161/cs161.1176/Sections/final_review-1.pdf

In practice this algorithm would be fine since the data size is small, but worth looking at that sorting-based one. It may even make the code simpler and shorter!

Reply:
I think this does not work, as what we actually want is to check every possible pair and see if they don't overlap to count it as possible xhr performance entry. However, the idea of the suggested algorithm is a way to make a non overlapping array but we don't actually want to make the entries not overlapping as this might return the wrong entry associated to the XHR.